python/tvm/relay/interpreter.py

```diff
@@ -83,48 +87,122 @@ def _arg_to_ast(arg):
+        of expression.
+
+        Parameters
+        ----------
+        expr: relay.Expr
+            The Relay expression to execute.
+
+        Returns
+        -------
+        executor: function
+            A Python function which implements the behavior of `expr`.
+        """
+        raise Exception("abstract method: please implement me.")
 
-    Parameters
-    ----------
-    env: tvm.relay.Environment
-        The global environment used.
+    def evaluate(self, expr, params=None):
+        """
+        Evaluate a Relay expression on the interpreter.
+
+        Parameters
+        ----------
+        expr: tvm.relay.Expr
+            The expression to evaluate.
+        """
+        if params:
+            scope_builder = ScopeBuilder()
+            for key, value in params:
+                scope_builder.let(key, value)
+            scope_builder.ret(expr)
+            expr = scope_builder.get()
 
-    expr: tvm.relay.Expr
-        The expression to evaluate.
+        if isinstance(expr, Function):
+            assert not ir_pass.free_vars(expr)
 
-    args: list of tvm.relay.Expr
-        The arguments to apply to the expression, only works
-        if the expression has a function type.
+        return self._make_executor(expr)
 
-    Returns
-    -------
-    value: tvm.relay.eval.Value
-        The value produced by evaluating the expression.
+
+class Interpreter(Executor):
+    """
+    A wrapper around the Relay interpreter, implements the excecutor interface.
     """
-    # assert len(args) == 0
-    relay_args = []
-    for arg in args:
-        relay_args.append(_arg_to_ast(arg))
-
-    # TODO: We need to move this optimization code into the optimizer/pass manager
-    if isinstance(expr, GlobalVar):
-        func = env[expr]
-        func = apply_passes(func, env)
-        env._add(expr, func, True)
-        opt_expr = Call(expr, relay_args)
-        # import pdb; pdb.set_trace()
-        return _interpreter.evaluate(env, opt_expr)
+    def __init__(self, env=None):
+        Executor.__init__(self, env)
+
+    def _make_executor(self, expr):
+        def _interp_wrapper(*args):
+            relay_args = []
+            for arg in args:
+                relay_args.append(_arg_to_ast(arg))
+
+            if isinstance(expr, GlobalVar):
+                func = self.env[expr]
+                func = self.optimize(func)
+                self.env._add(expr, func, True)
+                opt_expr = Call(expr, relay_args)
+                return _interpreter.evaluate(self.env, opt_expr)
+            else:
+                call = Call(expr, relay_args)
+                opt_expr = self.optimize(call)
+                return _interpreter.evaluate(self.env, opt_expr)
+
+        return _interp_wrapper
+
+
+class GraphRuntime(Executor):
+    """A wrapper around the TVM graph runtime, implements the Executor interface."""
+    def __init__(self, env=None):
+        Executor.__init__(self, env)
+
+    def _make_executor(self, expr):
```

Comment:
I think I am a little confused for this executor here. 1) Do we want to put compile and execution together? 2) Are we assuming that there is only one single output?

Reply:
The general idea is the evaluator returns a Python closure which when later invoked will first optimize the program as much as possible then evaluate it. This is very much a JIT style approach, we could move the optimizations into two functions one before generating this
function and one after?

The use case for compiling after seeing the arguments is that we can further specialize the code. Say for example we have generic code which we want to later call on concrete inputs, we can use this information to generate better TVM code, or match to accelerate primitives. I'm happy to keep discussing the best ways to implement this kind of behavior, but feel that eventually we will want some JIT compiler mode, which does some optimization at evaluation time.